[CodeGen][StaticDataPartitioning]Place local-linkage global variables in hot or unlikely prefixed sections based on profile information

mingmingl-llvm opened the pull request:
In this PR, static-data-splitter pass finds out the local-linkage global variables in {`.rodata`, `.data.rel.ro`, `bss`, `.data`} sections by analyzing machine instruction operands, and aggregates their accesses from code across functions. 

A follow-up item is to analyze global variable initializers and count for access from data.
* This limitation is demonstrated by `bss2` and `data3` in `llvm/test/CodeGen/X86/global-variable-partition.ll`.

Some stats of static-data-splitter with this patch:

**section**|**bss**|**rodata**|**data**
:-----:|:-----:|:-----:|:-----:
hot-prefixed section coverage|99.75%|97.71%|91.30%
unlikely-prefixed section size percentage|67.94%|39.37%|63.10%

1. The coverage is defined as `#perf-sample-in-hot-prefixed <data> section / #perf-sample in <data.*> section` for each <data> section. 
   * The perf command samples `MEM_INST_RETIRED.ALL_LOADS:u:pinned:precise=2` events at a high frequency (`perf -c 2251`) for 30 seconds. The profiled binary is built as non-PIE so `data.rel.ro` coverage data is not available.
2. The unlikely-prefixed `<data>` section size percentage is defined as `unlikely <data> section size / the sum size of <data>.* sections` for each `<data>` section

mingmingl-llvm committed:
[CodeGen][StaticDataPartitioning]Place module-internal global variables based on profile information

mingmingl-llvm committed:
add comment for bss22 and data3

mingmingl-llvm commented:
cc @Colibrow

mingmingl-llvm commented:
I also created https://github.com/llvm/llvm-project/pull/125758 and https://github.com/llvm/llvm-project/pull/125757 as two smaller splits of this PR.

petrhosek commented:
Is there a reason for using `.unlikely` as a prefix rather than `.cold`? `.cold` seems more obvious as an opposite of `.hot` and we use it elsewhere in LLVM, most notably hot-cold splitting.

Colibrow commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -92,63 +121,131 @@ bool StaticDataSplitter::splitJumpTablesWithProfiles(
-        // This is not a source block of jump table.
-        if (JTI == -1)
+        std::optional<uint64_t> Count = std::nullopt;
+        if (!Op.isJTI() && !Op.isGlobal())
           continue;
 
-        auto Hotness = MachineFunctionDataHotness::Hot;
+        Count = MBFI->getBlockProfileCount(&MBB);
```
```suggestion
	if (!Op.isJTI() && !Op.isGlobal())
          continue;

        std::optional<uint64_t> Count = MBFI->getBlockProfileCount(&MBB);
```
is that OK?

mingmingl-llvm replied:
Yes the change is ok.

I kept getting "Applying suggestions on deleted lines is currently not supported." error when trying to commit the suggestion. I'll make the change in the local client..

Colibrow commented on llvm/test/CodeGen/X86/data-section-prefix.ll:
```diff
@@ -0,0 +1,27 @@
+; COFF-MSVC: .section .bss,"bw",one_only,baz
+; COFF-MSVC: .section .bss,"bw",one_only,quz
+
+@foo = global i32 1, !section_prefix !0
+@bar = global i32 2
+@baz = global i32 0, !section_prefix !1
+@quz = global i32 0
+
+!0 = !{!"section_prefix", !"hot"}
+!1 = !{!"section_prefix", !"unlikely"}
```
I'm not sure if data_section_prefix would make sense.

mingmingl-llvm replied:
Function and global variables are derived classes of GlobalObject, and currently function section prefix metadata has `function_section_prefix`.

To me, `data_section_prefix` carries the semantic that a global variable/object is a piece of data, which may become inaccurate. Alternative to setting global object's section prefix, I can keep function's methods as they are in TOT, and implements `{set,get,update}SectionPrefix` for GlobalVariables (not its superclass GlobalObject) and name it `global_variable_section_prefix`. How does that sound?

Colibrow replied:
yes. I get it.

mingmingl-llvm reviewed (commented):
>  Is there a reason for using .unlikely as a prefix rather than .cold? .cold seems more obvious as an opposite of .hot and we use it elsewhere in LLVM, most notably hot-cold splitting.

@petrhosek Thanks for the question. `unlikely` is used to follow the conventions of [functions](https://github.com/llvm/llvm-project/blob/69ccb1357fa6cf72063c737d06d6b29ffc465bee/llvm/lib/CodeGen/CodeGenPrepare.cpp#L598) in the codegenprepare pass. I'm open to name changes.

If identical section names (as string literals) are not de-duplicated in the relocatable object files (e.g., when compiled with `-data-sections=true -unique-section-names=false`), `cold` will give smaller section names than `unlikely` per data section where this matters.

snehasish commented:
> > Is there a reason for using .unlikely as a prefix rather than .cold? .cold seems more obvious as an opposite of .hot and we use it elsewhere in LLVM, most notably hot-cold splitting.
> 
> @petrhosek Thanks for the question. `unlikely` is used to follow the conventions of [functions](https://github.com/llvm/llvm-project/blob/69ccb1357fa6cf72063c737d06d6b29ffc465bee/llvm/lib/CodeGen/CodeGenPrepare.cpp#L598) in the codegenprepare pass. I'm open to name changes.
> 
> If identical section names (as string literals) are not de-duplicated in the relocatable object files (e.g., when compiled with `-data-sections=true -unique-section-names=false`), `cold` will give smaller section names than `unlikely` per data section where this matters.

I believe the `.cold` suffix is used for symbol names whereas the `.unlikely` suffix is for section names. It would be inconsistent to have a `.text.unlikely` section and a `.rodata.cold` section but I don't have a particularly strong opinion.

mingmingl-llvm committed:
run 'git merge main'

mingmingl-llvm committed:
apply code review suggestions

mingmingl-llvm committed:
record global variable section prefix updates as module updates

williamweixiao commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -117,18 +142,89 @@ bool StaticDataSplitter::partitionStaticDataWithProfiles(MachineFunction &MF) {
 
+const GlobalVariable *
+StaticDataSplitter::getLocalLinkageGlobalVariable(const GlobalValue *GV) {
+  if (!GV || GV->isDeclarationForLinker())
```
Can "GV->hasLocalLinkage()" cover "GV->isDeclarationForLinker()"?

mingmingl-llvm replied:
Good question. I haven't thought about it before.

It turns out IR verifier [requires](https://github.com/llvm/llvm-project/blob/3e2afe5f019b7a1c60e23cb2743018bd2d0417b1/llvm/lib/IR/Verifier.cpp#L735-L736) a declaration to have one of 'valid' decl linkages, and valid decl linkages must be at least external according to [isValidDeclarationLinkage](https://github.com/llvm/llvm-project/blob/3e2afe5f019b7a1c60e23cb2743018bd2d0417b1/llvm/include/llvm/IR/GlobalValue.h#L420).

I removed `GV->isDeclarationForLinker()` here and added a comment.

williamweixiao reviewed (commented):
Looks good! Just some minor comments.

snehasish commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -117,18 +142,89 @@ bool StaticDataSplitter::partitionStaticDataWithProfiles(MachineFunction &MF) {
+    // framework, and set global variable section prefix once per module after
+    // analyzing all machine functions.
+    if (PSI->isColdCount(Iter->second.getZExtValue())) {
+      Changed |= GV.updateSectionPrefix("unlikely",
```
updateSectionPrefix returns void so is this UB? Regardless, I think we can drop the updateSectionPrefix helper in #125757.

mingmingl-llvm replied:
done.

snehasish commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -46,6 +49,20 @@ class StaticDataSplitter : public MachineFunctionPass {
+  const GlobalVariable *getLocalLinkageGlobalVariable(const GlobalValue *GV);
+
+  // Returns true if the global variable is in one of {.rodata, .bss, .data,
+  // .data.rel.ro} sections
```
nit: period at the end like the other comments?

snehasish commented on llvm/include/llvm/IR/GlobalObject.h:
```diff
@@ -124,6 +124,17 @@ class GlobalObject : public GlobalValue {
+
+  /// Update the section prefix, unless the existing prefix is the same as
+  /// `KeepPrefix`.
+  bool updateSectionPrefix(StringRef Prefix,
```
Don't forget to drop this after merging #125757.

snehasish commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -46,6 +49,20 @@ class StaticDataSplitter : public MachineFunctionPass {
+  bool updateGlobalVariableSectionPrefix(MachineFunction &MF);
+
+  // Accummulated data profile count across machine functions in the module.
+  DenseMap<const GlobalVariable *, APInt> DataProfileCounts;
```
Is there a strong reason to use APInt as the value? When passed to the `isColdCount` API we extract the value as a uint64_t anyway (with an assertion). I think if you are concerned about overflows then we can use `AddOverflow` from `Support/MathExtras.h` on L163 and add an assert. Additionally, there we could use a saturating count using `getInstrMaxCountValue` as the limit if we do have cases which overflow. Wdyt?

mingmingl-llvm replied:
I think it's a good idea to make counters `uint64_t` and clamp overflows as max. Done with saturating add and clamping to `getInstrMaxCountValue`. Also removed `#include "llvm/ADT/APInt.h"`

snehasish commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -117,18 +142,89 @@ bool StaticDataSplitter::partitionStaticDataWithProfiles(MachineFunction &MF) {
 
           if (MJTI->updateJumpTableEntryHotness(JTI, Hotness))
             ++NumChangedJumpTables;
+        } else if (Op.isGlobal()) {
```
Do we need to check `Op.isGlobal()` again? I think we can only get past L128 if Op is JTI or global and we check JTI on L133.

mingmingl-llvm replied:
It's not need for now (i.e., before `Op.isCPI` is used to handle constant pools) as you point out, so I removed it.

mingmingl-llvm commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -117,18 +142,89 @@ bool StaticDataSplitter::partitionStaticDataWithProfiles(MachineFunction &MF) {
+  assert(GV && "Caller guaranteed");
+
+  // Skip LLVM reserved symbols.
+  if (GV->getName().starts_with("llvm."))
```
Good question.  The globals with `llvm.` as a name prefix are usually handled specially, by many (middle-end and back-end) passes (e.g., AsmPrinter emits [special LLVM values](https://github.com/llvm/llvm-project/blob/7464dc8c7618aeb5a01998576bbcc4c88f0dde1d/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp#L730-L731) specially), and they are skipped in this pass mostly out of conservativeness. 

I moved this check before calling `inStaticDataSection` helper though for readability, and added a comment around the check. What do you think about this?

mingmingl-llvm commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -117,18 +142,89 @@ bool StaticDataSplitter::partitionStaticDataWithProfiles(MachineFunction &MF) {
+    // framework, and set global variable section prefix once per module after
+    // analyzing all machine functions.
+    if (PSI->isColdCount(Iter->second.getZExtValue())) {
+      Changed |= GV.updateSectionPrefix("unlikely",
```
My understanding is that functions have `.hot` or `.unlikely` prefixes when they are placed as a whole (in the [codegenprepare](https://github.com/llvm/llvm-project/blob/69ccb1357fa6cf72063c737d06d6b29ffc465bee/llvm/lib/CodeGen/CodeGenPrepare.cpp#L598) pass), and have `.split` as a prefix when basic blocks of a function are further categorized into hot and non-hot ones (in the [MFS](https://github.com/llvm/llvm-project/blob/69ccb1357fa6cf72063c737d06d6b29ffc465bee/llvm/lib/CodeGen/MachineFunctionSplitter.cpp#L12-L14) pass you mentioned).

Here the granularity is a piece of global variable, and `.split` would be a good name if a global variable itself is split into hot and cold ones. For instance, there should be hot and cold functions in a vtable, but it'd take substantial work to implement an ABI to support the split vtable entries.

mingmingl-llvm committed:
merge with main, and resolve code review comments

mingmingl-llvm reviewed (commented):
thanks for the reviews!

mingmingl-llvm committed:
remove blank line

mingmingl-llvm committed:
Implement module-wide analysis of global variable hotness.
* In StaticDataProfileInfo.h/cpp, add an immutable pass to keep track of
  constants and their profile information across functions in a module.
* Add a module pass, StaticDataAnnotator, to set global variable's
  section prefix based on module-wide hotness.

mingmingl-llvm committed:
run 'git merge main'

mingmingl-llvm committed:
port code de-duplication based on feedback in the follow up patch (https://github.com/llvm/llvm-project/pull/129781)

mingmingl-llvm commented:
Hi folks, just a quick note for this change when you have a moment :)

mingmingl-llvm commented on llvm/include/llvm/Analysis/StaticDataProfileInfo.h:
```diff
@@ -0,0 +1,68 @@
+  void addConstantProfileCount(const Constant *C,
+                               std::optional<uint64_t> Count);
+
+  /// If \p C has a count, return it. Otherwise, return std::nullopt.
```
I get the point that the difference between `hasUnknownCount` and `getConstantProfileCount` are subtle. I thought about this more.

Basically, StaticDataAnnotator uses both PSI and this class to decide the section prefix of a constant, so an alternative is (for this class) to provide `getConstantSectionPrefix` directly.  The updated change implements the alternative option, and makes `getConstantProfileCount` to a private class method.

mingmingl-llvm commented on llvm/include/llvm/Analysis/StaticDataProfileInfo.h:
```diff
@@ -0,0 +1,68 @@
+#ifndef LLVM_ANALYSIS_STATICDATAPROFILEINFO_H
```
Overall, I don't feel strong about having new classes in `ProfileSummaryInfo.h` or in new files. Actually I once thought about the former and hesitated between the two.

I ended up with the latter mainly because PSI is included in multiple cpp files, and the new classes don't need to be included as much.

As the static data partitioning work is under development and symbolized data access profiles may need helper functions and libraries, I think we can keep it this way and do refactors later where they fit. What do you think about this?

mingmingl-llvm commented on llvm/lib/CodeGen/StaticDataAnnotator.cpp:
```diff
@@ -0,0 +1,119 @@
+
+    // Skip global variables without profile counts. The module may not be
+    // profiled or instrumented.
+    auto Count = SDPI->getConstantProfileCount(&GV);
```
For the original code, the semantic of `getConstantProfileCount` and `hasUnknownCount` are subtle on the different combinations (i.e., whether a var is used in hot/cold/lukewarm/unprofiled functions, etc)

The updated code will query a section prefix from `StaticDataProfileInfo` directly there. PTAL.

mingmingl-llvm commented on llvm/lib/CodeGen/StaticDataAnnotator.cpp:
```diff
@@ -0,0 +1,119 @@
+      // prefix isn't hot already.
+      GV.setSectionPrefix("hot");
+      Changed = true;
+    } else if (PSI->isColdCount(*Count) && !SDPI->hasUnknownCount(&GV) &&
```
I think this question is similar to https://github.com/llvm/llvm-project/pull/125756/files#r2017471165 and https://github.com/llvm/llvm-project/pull/125756/files#r2017466522, and being discussed there.

mingmingl-llvm committed:
resolve comments

david-xl commented on llvm/lib/CodeGen/StaticDataAnnotator.cpp:
```diff
@@ -0,0 +1,103 @@
+      continue;
+
+    StringRef SectionPrefix = SDPI->getConstantSectionPrefix(&GV, PSI);
+    if (SectionPrefix.empty() || alreadyHasSectionPrefix(GV, SectionPrefix))
```
In what situation can the constant already have a section prefix?

snehasish replied:
clang provides a section attribute which can be attached to globals. This is commonly used in embedded applications as far as I know.


https://clang.llvm.org/docs/AttributeReference.html#section-declspec-allocate


@mingmingl-llvm I don't think we have a test which covers this condition behaviour?

mingmingl-llvm replied:
> In what situation can the constant already have a section prefix?

Good question. A search of `GlobalValue::setSectionPrefix` in the LLVM repo shows that it's only used for functions but not global variables, so the `alreadyHasSectionPrefix` should return always false currently. To be a little bit future proof, updated the patch to report error if a global variable already has a section prefix, and commented that the subsequent code doesn't handle 'update' properly.

> clang provides a section attribute which can be attached to globals. This is commonly used in embedded applications as far as I know.

Thanks for pointing this out. 

For test coverage purposes, the test case was updated with `@cold_data_custom_foo_section` and  `@hot_data_custom_bar_section`;  the former uses [clang's section attribute](https://clang.llvm.org/docs/AttributeReference.html#section-declspec-allocate), and the latter uses [#pragma clang section ](https://clang.llvm.org/docs/LanguageExtensions.html#specifying-section-names-for-global-objects-pragma-clang-section).

Taking a look at [1], such global variables [1] will have explicit sections, and this work (hot-cold partitioning) won't kick in. Per offline discussion, it's desired for compiler to keep source code annotations without partitioning.

[1] https://github.com/llvm/llvm-project/blob/91d2ecf0d563b03d75380375e8ac26a291bed9d7/llvm/lib/Target/TargetLoweringObjectFile.cpp#L316-L331

snehasish replied:
SGTM.

snehasish commented on llvm/lib/CodeGen/StaticDataAnnotator.cpp:
```diff
@@ -0,0 +1,103 @@
+// in the module, looks up counters from StaticDataProfileInfo and sets the
+// section prefix based on profiles.
+//
+// The three-pass structure is implemented for practical reasons, to work around
```
The text above could be a little bit clearer that StaticDataProfileInfoWrapper is also a pass. When reading three-pass here I was a little confused.

snehasish commented on llvm/lib/CodeGen/StaticDataAnnotator.cpp:
```diff
@@ -0,0 +1,103 @@
+
+using namespace llvm;
+
+class StaticDataAnnotator : public ModulePass {
```
Add a comment to describe this class.

snehasish commented on llvm/lib/CodeGen/StaticDataSplitter.cpp:
```diff
@@ -117,18 +164,37 @@ bool StaticDataSplitter::partitionStaticDataWithProfiles(MachineFunction &MF) {
+
+bool StaticDataSplitter::inStaticDataSection(const GlobalVariable *GV,
+                                             const TargetMachine &TM) {
+  assert(GV && "Caller guaranteed");
```
Change the param to a const & to avoid this assertion?

snehasish commented on llvm/test/CodeGen/X86/global-variable-partition.ll:
```diff
@@ -0,0 +1,198 @@
+
+; This RUN command sets `-data-sections=true -unique-section-names=true` so data
+; sections are uniqufied by numbers.
+; RUN: llc -mtriple=x86_64-unknown-linux-gnu -enable-split-machine-functions \
```
Reminder to update this test too once we remove the dependence on -enable-split-machine-functions.

snehasish reviewed (approved):
lgtm

mingmingl-llvm committed:
resolve comments

snehasish reviewed (approved):
lgtm

mingmingl-llvm commented:
The [presubmit failure](https://buildkite.com/llvm-project/github-pull-requests/builds/162943#0195de6c-2ce9-474c-b77d-012f0b08e177) in `LLVM.tools/llvm-exegesis/RISCV/rvv/filter.test` doesn't look relevant according to error message, and the test passed locally (with `-DLLVM_TARGETS_TO_BUILD="X86;AArch64;RISCV"`). So it might be flakiness.
